File: frontend/src/types/application.ts
/**
 * TypeScript types for FGN Savings Bond application.
 */

export type ApplicantType = 'Individual' | 'Joint' | 'Corporate';
export type Tenor = '2-Year' | '3-Year';

export interface ApplicationFormData {
  // Bond details
  tenor: Tenor;
  month_of_offer: string;
  bond_value: number;
  amount_in_words: string;

  // Applicant type
  applicant_type: ApplicantType;

  // Individual/Joint applicant fields (Primary)
  title?: string;
  full_name?: string;
  date_of_birth?: string;
  phone_number?: string;
  email?: string;
  occupation?: string;
  passport_no?: string;
  next_of_kin?: string;
  mothers_maiden_name?: string;
  address?: string;
  cscs_number?: string;
  chn_number?: string;

  // Joint applicant fields (Secondary)
  joint_title?: string;
  joint_full_name?: string;
  joint_date_of_birth?: string;
  joint_phone_number?: string;
  joint_email?: string;
  joint_occupation?: string;
  joint_passport_no?: string;
  joint_next_of_kin?: string;
  joint_address?: string;

  // Corporate fields
  company_name?: string;
  rc_number?: string;
  business_type?: string;
  contact_person?: string;
  corp_phone_number?: string;
  corp_email?: string;
  corp_passport_no?: string;

  // Bank details (Primary)
  bank_name: string;
  bank_branch?: string;
  account_number: string;
  sort_code?: string;
  bvn?: string;

  // Joint bank details
  joint_bank_name?: string;
  joint_bank_branch?: string;
  joint_account_number?: string;
  joint_sort_code?: string;
  joint_bvn?: string;

  // Classification
  is_resident: boolean;
  investor_category?: string[];

  // Distribution agent
  agent_name?: string;
  stockbroker_code?: string;

  // Witness
  needs_witness: boolean;
  witness_name?: string;
  witness_address?: string;
  witness_acknowledged: boolean;
}

export interface ApplicationResponse extends ApplicationFormData {
  id: number;
  submission_date: string;
  created_at?: string;
  payment_status: PaymentStatus;
}

// Payment types
export type PaymentStatus = 'pending' | 'paid' | 'verified' | 'rejected';
export type PaymentMethod = 'bank_transfer' | 'cheque' | 'cash' | 'pos' | 'other';

export interface PaymentCreate {
  amount: number;
  payment_method: PaymentMethod;
  payment_reference: string;
  payment_date: string;
  receiving_bank?: string;
  notes?: string;
}

export interface PaymentUpdate {
  amount?: number;
  payment_method?: PaymentMethod;
  payment_reference?: string;
  payment_date?: string;
  receiving_bank?: string;
  notes?: string;
}

export interface PaymentVerify {
  action: 'verify' | 'reject';
  rejection_reason?: string;
}

export interface PaymentDocument {
  id: number;
  payment_id: number;
  filename: string;
  original_filename: string;
  file_size: number | null;
  mime_type: string | null;
  uploaded_at: string;
}

export interface Payment {
  id: number;
  application_id: number;
  amount: number;
  payment_method: string;
  payment_reference: string;
  payment_date: string;
  receiving_bank: string | null;
  status: PaymentStatus;
  verified_at: string | null;
  verified_by: string | null;
  rejection_reason: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string | null;
  documents: PaymentDocument[];
}

// DMO Reporting types
export interface MonthlyReportSummary {
  month_of_offer: string;
  year: number;
  total_applications: number;
  total_value: number;
  average_value: number;
  total_2year: number;
  value_2year: number;
  total_3year: number;
  value_3year: number;
  total_individual: number;
  total_joint: number;
  total_corporate: number;
  pending_count: number;
  pending_value: number;
  paid_count: number;
  paid_value: number;
  verified_count: number;
  verified_value: number;
  rejected_count: number;
  rejected_value: number;
  is_submitted: boolean;
  submission_id: number | null;
  submitted_at: string | null;
}

export interface DMOSubmission {
  id: number;
  month_of_offer: string;
  year: number;
  total_applications: number;
  total_value: number;
  total_2year: number;
  total_3year: number;
  total_verified: number;
  submitted_at: string;
  submitted_by: string | null;
  report_file_path: string | null;
  notes: string | null;
}

export interface DMOSubmissionCreate {
  month_of_offer: string;
  year: number;
  notes?: string;
}

export interface Constants {
  banks: string[];
  investor_categories: string[];
  months: string[];
  tenors: string[];
  titles: string[];
}

// Form wizard step definition
export interface WizardStep {
  id: number;
  title: string;
  description: string;
}

export const WIZARD_STEPS: WizardStep[] = [
  { id: 0, title: 'Bond Details', description: 'Select tenor and enter bond value' },
  { id: 1, title: 'Applicant Type', description: 'Choose applicant category' },
  { id: 2, title: 'Applicant Info', description: 'Enter personal/company details' },
  { id: 3, title: 'Bank Details', description: 'Enter bank account information' },
  { id: 4, title: 'Classification', description: 'Select residency and investor category' },
  { id: 5, title: 'Distribution', description: 'Enter agent details (optional)' },
  { id: 6, title: 'Review & Submit', description: 'Review and submit application' },
];

File: backend/app/models/application.py
"""SQLAlchemy model for bond subscription applications."""

from datetime import datetime

from sqlalchemy import CheckConstraint, Column, ForeignKey, Index, Integer, String, Text
from sqlalchemy.orm import relationship

from ..database import Base


class Application(Base):
    """
    FGN Savings Bond subscription application model.

    Stores all ~70 fields for Individual, Joint, and Corporate applicant types.
    """

    __tablename__ = "applications"

    # Primary key
    id = Column(Integer, primary_key=True, index=True)

    # Submission metadata
    submission_date = Column(String(30), nullable=False)
    created_at = Column(String(30), default=lambda: datetime.now().isoformat())

    # Bond details
    tenor = Column(String(10), nullable=False)  # "2-Year" or "3-Year"
    month_of_offer = Column(String(20), nullable=False)
    bond_value = Column(Integer, nullable=False)  # Store as kobo/integer for precision
    amount_in_words = Column(Text, nullable=False)

    # Applicant type
    applicant_type = Column(String(20), nullable=False)  # Individual, Joint, Corporate

    # Individual/Joint applicant fields (Primary applicant)
    title = Column(String(20))
    full_name = Column(String(200))
    date_of_birth = Column(String(20))
    phone_number = Column(String(20))
    email = Column(String(100))
    occupation = Column(String(100))
    passport_no = Column(String(50))
    next_of_kin = Column(String(200))
    mothers_maiden_name = Column(String(200))
    address = Column(Text)
    cscs_number = Column(String(20))
    chn_number = Column(String(20))

    # Joint applicant fields (Secondary applicant)
    joint_title = Column(String(20))
    joint_full_name = Column(String(200))
    joint_date_of_birth = Column(String(20))
    joint_phone_number = Column(String(20))
    joint_email = Column(String(100))
    joint_occupation = Column(String(100))
    joint_passport_no = Column(String(50))
    joint_next_of_kin = Column(String(200))
    joint_address = Column(Text)

    # Corporate applicant fields
    company_name = Column(String(200))
    rc_number = Column(String(50))
    business_type = Column(String(100))
    contact_person = Column(String(200))
    corp_phone_number = Column(String(20))
    corp_email = Column(String(100))
    corp_passport_no = Column(String(50))

    # Bank details (Primary applicant)
    bank_name = Column(String(100), nullable=False)
    bank_branch = Column(String(200))
    account_number = Column(String(10), nullable=False)
    sort_code = Column(String(20))
    bvn = Column(String(11))

    # Joint applicant bank details
    joint_bank_name = Column(String(100))
    joint_bank_branch = Column(String(200))
    joint_account_number = Column(String(10))
    joint_sort_code = Column(String(20))
    joint_bvn = Column(String(11))

    # Classification
    is_resident = Column(Integer, nullable=False, default=1)  # 1=Resident, 0=Non-Resident
    investor_category = Column(Text)  # JSON array stored as text

    # Distribution agent
    agent_name = Column(String(200))
    stockbroker_code = Column(String(50))

    # Witness section (for illiterate applicants)
    needs_witness = Column(Integer, default=0)  # 0=No, 1=Yes
    witness_name = Column(String(200))
    witness_address = Column(Text)
    witness_acknowledged = Column(Integer, default=0)

    # Payment tracking
    payment_status = Column(String(20), default="pending")  # pending, paid, verified, rejected
    dmo_submission_id = Column(
        Integer,
        ForeignKey("dmo_submissions.id", ondelete="SET NULL"),
        nullable=True,
    )

    # Relationships
    payment = relationship(
        "Payment",
        back_populates="application",
        uselist=False,  # One-to-one relationship
        cascade="all, delete-orphan",
    )
    dmo_submission = relationship("DMOSubmission", back_populates="applications")

    # Table constraints
    __table_args__ = (
        CheckConstraint("tenor IN ('2-Year', '3-Year')", name="valid_tenor"),
        CheckConstraint(
            "applicant_type IN ('Individual', 'Joint', 'Corporate')",
            name="valid_applicant_type",
        ),
        CheckConstraint(
            "bond_value >= 5000 AND bond_value <= 50000000", name="valid_bond_value"
        ),
        CheckConstraint("length(account_number) = 10", name="valid_account_number"),
        CheckConstraint(
            "bvn IS NULL OR length(bvn) = 11", name="valid_bvn"
        ),
        CheckConstraint("is_resident IN (0, 1)", name="valid_is_resident"),
        CheckConstraint(
            "payment_status IN ('pending', 'paid', 'verified', 'rejected')",
            name="valid_payment_status",
        ),
        # Indexes for common queries
        Index("idx_applicant_type", "applicant_type"),
        Index("idx_submission_date", "submission_date"),
        Index("idx_month_of_offer", "month_of_offer"),
        Index("idx_tenor", "tenor"),
        Index("idx_payment_status", "payment_status"),
    )

    def __repr__(self) -> str:
        if self.applicant_type == "Corporate":
            name = self.company_name
        else:
            name = self.full_name
        return f"<Application(id={self.id}, type={self.applicant_type}, name={name})>"
